File: aula03-analise-bigo/python/ex03_busca_binaria.py
def busca_binaria(lista, alvo):
    esquerda, direita = 0, len(lista) - 1
    while esquerda <= direita:
        # divisao inteira pra nao estourar indice
        meio = (esquerda + direita) // 2
        if lista[meio] == alvo:
            return meio
        elif lista[meio] < alvo:
            esquerda = meio + 1
        else:
            direita = meio - 1
    return -1


if __name__ == "__main__":
    ordenada = [1, 3, 5, 7, 9, 11, 13]
    print("Indice do 7:", busca_binaria(ordenada, 7))
    print("Indice do 8 (ausente):", busca_binaria(ordenada, 8))

File: aula04-reescrita-rust/src/ex03_busca_binaria.rs
pub fn busca_binaria(lista: &[i32], alvo: i32) -> Option<usize> {
    let mut esquerda: isize = 0;
    // isize porque direita pode chegar a -1
    let mut direita: isize = lista.len() as isize - 1;

    while esquerda <= direita {
        let meio: isize = (esquerda + direita) / 2;
        let valor = lista[meio as usize];

        if valor == alvo {
            return Some(meio as usize);
        } else if valor < alvo {
            esquerda = meio + 1;
        } else {
            direita = meio - 1;
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encontra_elemento_existente() {
        let v = vec![1, 3, 5, 7, 9, 11];
        assert_eq!(busca_binaria(&v, 7), Some(3));
    }

    #[test]
    fn encontra_primeiro_e_ultimo() {
        let v = vec![1, 3, 5, 7, 9, 11];
        assert_eq!(busca_binaria(&v, 1), Some(0));
        assert_eq!(busca_binaria(&v, 11), Some(5));
    }

    #[test]
    fn nao_encontra_retorna_none() {
        let v = vec![1, 3, 5, 7, 9, 11];
        assert_eq!(busca_binaria(&v, 4), None);
        assert_eq!(busca_binaria(&v, 0), None);
        assert_eq!(busca_binaria(&v, 99), None);
    }

    #[test]
    fn lista_vazia_retorna_none() {
        let v: Vec<i32> = Vec::new();
        assert_eq!(busca_binaria(&v, 1), None);
    }
}
